### PR TITLE
Output particle data as binary file

### PR DESCRIPTION
**Pull request by koarakawaii**
1. Change `OPT__OUTPUT_PAR_TEXT` to `OPT__OUTPUT_PAR_MODE`.
2. Add `Particle/Par_Output_BinaryFile.cpp`
3. Modify headers and functions in the source code accordingly,
4. Set OPT__OUTPUT_PAR_MODE to 2 in `Input__Parameter` (0=off, 1=text-file, 2=C-binary, default 0)
5. Data format
    * Total data numbers: `N_attribute*N_activate_particle`
    * Data of all activate particles with a selected attribute will be dumped consecutively, followed by the next attribute, so on and so forth

### CODE REVIEW

File: src/Output/Output_DumpData_Total_HDF5.cpp [hunk 1]
```diff
@@ -2564,7 +2564,7 @@ void FillIn_InputPara( InputPara_t &InputPara, const int NFieldStored, char Fiel
    InputPara.Opt__Output_Part        = OPT__OUTPUT_PART;
    InputPara.Opt__Output_User        = OPT__OUTPUT_USER;
 #  ifdef PARTICLE
-   InputPara.Opt__Output_ParText     = OPT__OUTPUT_PAR_TEXT;
+   InputPara.Opt__Output_Par_Mode     = OPT__OUTPUT_PAR_MODE;
```
Comment (hyschive) [inline src/Output/Output_DumpData_Total_HDF5.cpp:2567]: Code alignment

File: tool/analysis/gamer_extract_uniform/Makefile [hunk 2]
```diff
@@ -10,7 +10,7 @@ EXECUTABLE  = GAMER_ExtractUniform
 # siimulation options
 #######################################################################################################
 # model: HYDRO/MHD/ELBDM
-SIMU_OPTION += -DMODEL=HYDRO
+SIMU_OPTION += -DMODEL=ELBDM
```
Comment (hyschive) [inline tool/analysis/gamer_extract_uniform/Makefile:13]: This file should not be modified by this PR.

File: tool/inits/GEN_UM_IC_WITH_PHASE_WITH_TRANSFER_FUNC/work_note_backup [hunk 3]
```diff
@@ -0,0 +1,16 @@
+Issue related to CAMB spectrum:
+    CAMB evolves the spectrum of each k component independently, so all compoenents might not grow with an universal growth factor. We observed that only those k modes in the intermedium regime scales according to the growth factor of linear-theorem (d_plus), while for very high k and very low k, the linear theorem fails. The high k might be due to the non-linear effect, while the low k mode we haven't have a explanation yet.
```
Comment (hyschive) [inline tool/inits/GEN_UM_IC_WITH_PHASE_WITH_TRANSFER_FUNC/work_note_backup:2]: This does not belong to this PR.

File: src/Particle/Par_Output_BinaryFile.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
+#include "GAMER.h"
+
+#ifdef PARTICLE
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Par_Output_BinaryFile
+// Description :  Output the particle position and velocity
```
Comment (hyschive) [inline src/Particle/Par_Output_BinaryFile.cpp:6]: Clarify the binary data format

File: src/Particle/Par_Output_BinaryFile.cpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+//
+// Parameter   :  FileName : Output file name
+//
+// Return      :  None
+//-------------------------------------------------------------------------------------------------------
+void Par_Output_BinaryFile( const char *FileName )
+{
+   if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s (DumpID = %d) ...\n", __FUNCTION__, DumpID );
+//   int N_att = PAR_NATT_TOTAL;
```
Comment (hyschive) [inline src/Particle/Par_Output_BinaryFile.cpp:15]: Remove it.

File: src/Particle/Par_Output_BinaryFile.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+//                MPI_Gather(&counter, 1, MPI_LONG, &counter_gather, 1, MPI_LONG, 0, MPI_COMM_WORLD);
+//                counter_gather += counter;
+//                if (MPI_Rank==0) 
+//                {
+//                    fwrite(&counter_gather, sizeof(long), 1, File);
+//                    printf("%ld\n", counter_gather);
+//                    fwrite(&N_att, sizeof(int), 1, File);
+//                }
+//            }
+////
```
Comment (hyschive) [inline src/Particle/Par_Output_BinaryFile.cpp:60]: Remove these lines if they are not used anymore.

File: src/Particle/Par_Output_BinaryFile.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+{
+   if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s (DumpID = %d) ...\n", __FUNCTION__, DumpID );
+//   int N_att = PAR_NATT_TOTAL;
+   FILE *File;
+
+// check
+   if ( MPI_Rank == 0  &&  Aux_CheckFileExist(FileName) )
+   {
+      Aux_Message( stderr, "WARNING : file \"%s\" already exists and will be overwritten !!\n", FileName );
+      File = fopen( FileName, "w" );
```
Comment (hyschive) [inline src/Particle/Par_Output_BinaryFile.cpp:22]: Use `"wb"` for compatibility.

File: src/Particle/Par_Output_BinaryFile.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+   real *attribute_buff = (real*)malloc(sizeof(real)*amr->Par->NPar_AcPlusInac);
+// data
+   for (int v=0; v<PAR_NATT_TOTAL; v++)
+   {
+      for (int TargetMPIRank=0; TargetMPIRank<MPI_NRank; TargetMPIRank++)
+      {
+         if ( MPI_Rank == TargetMPIRank )
+         {
+            File = fopen( FileName, "a" );
```
Comment (hyschive) [inline src/Particle/Par_Output_BinaryFile.cpp:35]: Use `"ab"` for compatibility.

File: src/Output/Output_DumpData.cpp [hunk 9]
```diff
@@ -137,8 +137,10 @@ void Output_DumpData( const int Stage )
       sprintf( FileName_PS, "PowerSpec_%06d", DumpID );
 
 #  ifdef PARTICLE
-   if ( OPT__OUTPUT_PAR_TEXT )
+   if ( OPT__OUTPUT_MODE==1 )
       sprintf( FileName_Particle, "Particle_%06d", DumpID );
+   if ( OPT__OUTPUT_MODE==2 )
+      sprintf( FileName_Particle, "Particle_%06d.cbin", DumpID );
```
Comment (hyschive) [inline src/Output/Output_DumpData.cpp:143]: - It should be `OPT__OUTPUT_PAR_MODE` instead of `OPT__OUTPUT_MODE`.
- How about using `"Particle_%06d.txt"` to distinguish it from `"Particle_%06d.cbin"`? Note that it will break the compatibility with some old gnuplot plotting scripts (e.g., `example/test_problem/Hydro/ClusterMerger/gnuplot_script/plot__particle_single-cluster.gpt` and `example/test_problem/Hydro/Plummer/plot_script/plot_particles_*.gpt`).
- Add one space before and after `==` to conform to the GAMER format.

**Review by hyschive — CHANGES_REQUESTED** [review]
@koarakawaii Thanks for the contribution! Please check both the inline comments and the additional comments below.
 
- [ ] Replace `OPT__OUTPUT_PAR_TEXT` with `OPT__OUTPUT_PAR_MODE` and the corresponding description in `example/test_problem/Hydro/*/Input__Parameter` and `example/input/Input__Parameter`.
- [ ] Add the `typedef` variables in `include/Typedef.h` to map to `OPT__OUTPUT_PAR_MODE=1/2`.  See this [example](https://github.com/gamer-project/gamer/blob/master/include/Typedef.h#L133).
- [ ] Try to improve further the conformation to the GAMER format in `Par_Output_BinaryFile.cpp` (e.g., remove trailing spaces, add extra spaces within parentheses, add extra empty lines to better separate different code sections).

**Comment by koarakawaii** [comment]
@hyschive Thank you for the suggestion and correction. I have re-requested a review for the change made accordingly to the conversations above. Please have a check~ (I also modify the gnuplot scripts to make them compatible with the modified particle text file name)

**Review by hyschive — APPROVED** [review]

**Comment by hyschive** [comment]
@koarakawaii Thanks for the contribution!
